server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -262,6 +279,48 @@ protected void doClose() {
         }
     }
 
+    // Inspects all cluster state elements that contain a hint about what the current repository generation is and updates
+    // #latestKnownRepoGen if a newer than currently known generation is found
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        final String repoName = metadata.name();
+        long bestGenerationFromCS = RepositoryData.EMPTY_REPO_GEN;
+
+        final SnapshotsInProgress snapshotsInProgress = state.custom(SnapshotsInProgress.TYPE);
+        if (snapshotsInProgress != null) {
+            final SnapshotsInProgress.Entry entry = snapshotsInProgress.entries().stream()
+                .filter(e -> e.snapshot().getRepository().equals(repoName)).findFirst()
+                .orElse(null);
+            if (entry != null) {
+                bestGenerationFromCS = entry.getRepositoryStateId();
+            }
+        }
+
+        final SnapshotDeletionsInProgress deletionsInProgress = state.custom(SnapshotDeletionsInProgress.TYPE);
+        if (bestGenerationFromCS == RepositoryData.EMPTY_REPO_GEN && deletionsInProgress != null) {
```

Comment:
let's remove this extra condition (`bestGenerationFromCS == RepositoryData.EMPTY_REPO_GEN`), it's an optimization which does not mattter I think and could hurt us later.

Reply:
No actually that's not true I think. We have the sitaution where if you abort a snapshot, the delete entry will be created with `currentRepoGeneration + 1` since that's what the repo will be at when the delete actually runs. That's why I added this.

Reply:
I don't understand this condition. Why are we not just taking the max of all the entries that we see? Also, I suppose that we don't allow concurrent operations right now, so we assume that we have only one of these 3 metadata for the current repo?

Reply:
There is the one special case of aborting a snapshot where you can have an in progress snapshot and a delete. The delete will contain the future repository generation for after the snapshot finished -> we can't use that. Thats why I added this condition.

Reply:
Can we fix that situation? And can you add a comment to that effect? The condition as it stands here today is very unintuitive.

Reply:
> Can we fix that situation?

Yes, once we have #49060 we can clean this up when implementing concurrent operations on the repo. That will require the whole business of associating an operation with a strict repo generation to go away anyway.

=> Added a comment for now :)